tests/linalg_test.py

```diff
@@ -1712,17 +1712,38 @@ def osp_fun(A):
   @jtu.sample_product(
     # Skip empty shapes because scipy fails: https://github.com/scipy/scipy/issues/1532
     shape=[(3, 4), (3, 3), (4, 3)],
-    dtype=[np.float32],
+    dtype=float_types + complex_types,
     mode=["full", "r", "economic"],
+    pivoting=[False, True]
   )
-  def testScipyQrModes(self, shape, dtype, mode):
+  def testScipyQrModes(self, shape, dtype, mode, pivoting):
+    is_not_cpu_test_device = not jtu.test_device_matches(["cpu"])
+    is_not_valid_jaxlib_version = jtu.jaxlib_version() <= (0, 4, 38)
+    if pivoting and (is_not_cpu_test_device or is_not_valid_jaxlib_version):
+      self.skipTest("Pivoting is only supported on CPU with jaxlib > 0.4.38")
     rng = jtu.rand_default(self.rng())
-    jsp_func = partial(jax.scipy.linalg.qr, mode=mode)
-    sp_func = partial(scipy.linalg.qr, mode=mode)
+    jsp_func = partial(jax.scipy.linalg.qr, mode=mode, pivoting=pivoting)
+    sp_func = partial(scipy.linalg.qr, mode=mode, pivoting=pivoting)
     args_maker = lambda: [rng(shape, dtype)]
     self._CheckAgainstNumpy(sp_func, jsp_func, args_maker, rtol=1E-5, atol=1E-5)
     self._CompileAndCheck(jsp_func, args_maker)
 
+    # Pivoting is unsupported by the numpy api - repeat the jvp checks performed
+    # in NumpyLinalgTest::testQR for the `pivoting=True` modes here.
+    def qr_and_mul(a):
+      if mode == "r":
+        r, *_ = jsp_func(a)
+        q = r
+      else:
+        q, r, *_ = jsp_func(a)
+      return q @ r
+
+    m, n = shape
+    full_matrices = mode in ["full", "r"]
```

Comment:
I'm still seeing failures here when `mode == "r"`. Perhaps we should just test for `mode == "full"`? Regardless, we might also want to "undo" the pivoting in `qr_and_mul` so that any changes in the pivoting caused by the finite difference used in the check doesn't lead to unexpected results. What do you think?

Reply:
The `mode=="full"` case should be sufficient on its own, but we can easily include the `mode=="economic"` case as well so I'm happy to just ignore `mode=="r"` (this is implicitly tested in the other modes anyway). I'm not sure about the pivoting, but I think you are probably right - I will think about it and let you know later.

Reply:
I've updated `qr_and_mul` to "undo" the pivoting - it now expresses the identity function, like in the numpy tests and should be more robust.